File: picture-server/src/main/java/com/applory/pictureserver/domain/request/RequestDto.java
package com.applory.pictureserver.domain.request;

import com.applory.pictureserver.shared.Constant;
import com.fasterxml.jackson.annotation.JsonFormat;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.format.annotation.DateTimeFormat;

import javax.validation.constraints.NotNull;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

public class RequestDto {

    @Getter
    @Setter
    public static class Create {
        @NotNull
        private Constant.Specialty specialty;

        @NotNull
        private String title;

        @NotNull
        private Integer desiredPrice;

        @NotNull
        private LocalDateTime dueDate;

        @NotNull
        private String description;

        private String matchYn;

        private String completeYn;
    }

    @Getter
    @Setter
    @NoArgsConstructor
    public static class VM {
        private String id;
        private String userId;
        private String userProfileFileName;
        private String userNickname;
        private Double userAcceptRate;
        private Constant.Specialty specialty;
        private String title;
        private Integer desiredPrice;
        @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd HH:mm:ss", timezone = "Asia/Seoul")
        private LocalDateTime dueDate;
        private String description;
        private String matchYn;
        private Integer readCount;
        private Integer chatCount;
        private List<RequestDto.VM> anotherRequests;

        public VM(Request request) {
            this.id = request.getId();
            this.userId = request.getUser().getId();
            this.userNickname = request.getUser().getNickname();
            this.specialty = request.getSpecialty();
            this.title = request.getTitle();
            this.desiredPrice = request.getDesiredPrice();
            this.dueDate = request.getDueDate();
            this.description = request.getDescription();
            this.readCount = request.getReadCount();
            this.matchYn = request.getMatchYN();
            this.chatCount = request.getChatCount();

            if (Objects.nonNull(request.getUser().getFile())) {
                this.userProfileFileName = request.getUser().getFile().getStoreFileName();
            }
        }
    }

    @Getter
    @Setter
    public static class Search {
        private Constant.Specialty specialty;


        @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
        private LocalDateTime fromForDueDt;

        @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
        private LocalDateTime toForDueDt;

        private String userId;

        private String exceptThisId;

    }


}

File: picture-server-kt/src/main/kotlin/com/applory/pictureserverkt/request/RequestDto.kt
package com.applory.pictureserverkt.request

import org.jetbrains.annotations.NotNull
import org.springframework.format.annotation.DateTimeFormat
import java.time.LocalDateTime
import java.util.*
import javax.validation.constraints.NotEmpty

class RequestDto {

    companion object {
        @JvmStatic
        fun convertToVM(request: Request): VM {
            return VM(
                id = request.id!!,
                userId = request.user!!.id!!,
                userNickname = request.user!!.nickname!!,
                requestType = request.requestType!!,
                title = request.title!!,
                description = request.description!!,
                desiredPrice = request.desiredPrice!!,
                dueDate = request.dueDate!!,
                matchYn = request.matchYN!!,
                readCount = request.readCount!!
            )
        }
    }

    data class VM(
        val id: UUID,
        val userId: UUID,
        val userNickname: String,
        val requestType: Request.RequestType,
        val title: String,
        val description: String,
        val desiredPrice: Int,
        val dueDate: LocalDateTime,
        val matchYn: String,
        val readCount: Int,
        val chatCount: Int? = null,
        var anotherRequests: List<VM>? = null,
        var userAcceptRate: Double? = null
    )

    data class Create(
        @field:NotNull
        var requestType: Request.RequestType,

        @field:NotEmpty
        var title: String?,

        @field:NotNull
        var desiredPrice: Int,

        @field:NotNull
        var dueDate: LocalDateTime,

        @field:NotEmpty
        var description: String,

        var matchYN: String? = null,

        var completeYN: String? = null
    )

    data class Search(
        val requestType: Request.RequestType? = null,

        val userId: UUID? = null,

        val exceptThisId: UUID? = null,

        @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
        val fromForDueDt: LocalDateTime? = null,

        @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
        val toForDueDt: LocalDateTime? = null
    )


}
